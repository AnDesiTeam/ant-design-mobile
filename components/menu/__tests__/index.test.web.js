import React from 'react';
import { render, mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import Menu from '../index.web';

describe('Menu', () => {
  const data = [
    {
      label: 'Item1',
      value: '1',
    },
    {
      label: 'Item2',
      value: '2',
      children: [
        {
          label: 'Item21',
          value: '21',
        },
        {
          label: 'Item22',
          value: '22',
        },
      ],
    },
  ];

  it('renders correctly', () => {
    const wrapper = render(<Menu data={data} />);
    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('controll by value', () => {
    const wrapper = mount(<Menu data={data} value={['2', '22']} />);
    expect(wrapper.find('.am-list-item').at(1).hasClass('am-menu-selected')).toBe(true);
    expect(wrapper.find('.am-list-item').at(3).hasClass('am-sub-menu-item-selected')).toBe(true);
  });

  it('fires change event when item is selected', () => {
    jest.useFakeTimers();
    const handleChange = jest.fn();
    const wrapper = mount(<Menu data={data} onChange={handleChange} />);
    wrapper.find('.am-list-item').at(1).simulate('click');
    wrapper.find('.am-list-item').at(3).find('.am-radio-input').simulate('change');
    jest.runAllTimers();
    expect(handleChange).toBeCalledWith(['2', '22']);
  });

  describe('level 1', () => {
    it('renders correctly', () => {
      const wrapper = render(<Menu data={data} level="1" />);
      expect(renderToJson(wrapper)).toMatchSnapshot();
    });

    it('controll by value', () => {
      const wrapper = mount(<Menu data={data} level="1" value={['2']} />);
      expect(wrapper.find('.am-list-item').at(1).hasClass('am-sub-menu-item-selected')).toBe(true);
    });

    it('fires change event when item is selected', () => {
      jest.useFakeTimers();
      const handleChange = jest.fn();
      const wrapper = mount(<Menu data={data} level="1" onChange={handleChange} />);
      wrapper.find('.am-list-item').at(1).find('.am-radio-input').simulate('change');
      jest.runAllTimers();
      expect(handleChange).toBeCalledWith(['2']);
    });
  });

  describe('multiple select menu', () => {
    it('renders correctly', () => {
      const wrapper = render(<Menu data={data} multiSelect />);
      expect(renderToJson(wrapper)).toMatchSnapshot();
    });

    it('controll by value', () => {
      const wrapper = mount(<Menu data={data} multiSelect value={['2', ['21', '22']]} />);
      expect(wrapper.find('.am-list-item').at(1).hasClass('am-menu-selected')).toBe(true);
      expect(wrapper.find('.am-list-item').at(2).hasClass('am-sub-menu-item-selected')).toBe(true);
      expect(wrapper.find('.am-list-item').at(3).hasClass('am-sub-menu-item-selected')).toBe(true);
    });

    it('fires change event when item is selected', () => {
      jest.useFakeTimers();
      const handleChange = jest.fn();
      const wrapper = mount(<Menu data={data} multiSelect onChange={handleChange} />);
      wrapper.find('.am-list-item').at(1).simulate('click');
      wrapper.find('.am-list-item').at(2).find('.am-checkbox-input').simulate('change');
      wrapper.find('.am-list-item').at(3).find('.am-checkbox-input').simulate('change');
      jest.runAllTimers();
      expect(handleChange).toHaveBeenLastCalledWith(['2', ['21', '22']]);
    });
  });

  describe('multiple select menu level 1', () => {
    it('renders correctly', () => {
      const wrapper = render(<Menu data={data} multiSelect level="1" />);
      expect(renderToJson(wrapper)).toMatchSnapshot();
    });

    it('controll by value', () => {
      const wrapper = mount(<Menu data={data} multiSelect level="1" value={['1', '2']} />);
      expect(wrapper.find('.am-list-item').at(0).hasClass('am-sub-menu-item-selected')).toBe(true);
      expect(wrapper.find('.am-list-item').at(1).hasClass('am-sub-menu-item-selected')).toBe(true);
    });

    it('fires change event when item is selected', () => {
      jest.useFakeTimers();
      const handleChange = jest.fn();
      const wrapper = mount(<Menu data={data} multiSelect level="1" onChange={handleChange} />);
      wrapper.find('.am-list-item').at(0).find('.am-checkbox-input').simulate('change');
      wrapper.find('.am-list-item').at(1).find('.am-checkbox-input').simulate('change');
      jest.runAllTimers();
      expect(handleChange).toHaveBeenLastCalledWith(['1', '2']);
    });

    it('fires onOk event when ok btn is clicked', () => {
      const onOk = jest.fn();
      const wrapper = mount(<Menu data={data} multiSelect value={['1', '2']} level="1" onOk={onOk} />);
      wrapper.find('.am-multi-select-btns-btn').at(1).simulate('click');
      expect(onOk).toBeCalledWith(['1', '2']);
    });

    it('fires onCancel event when cancel btn is clicked', () => {
      const onCancel = jest.fn();
      const wrapper = mount(<Menu data={data} multiSelect value={['1', '2']} level="1" onCancel={onCancel} />);
      wrapper.find('.am-multi-select-btns-btn').at(0).simulate('click');
      expect(onCancel).toHaveBeenCalled();
    });
  });
});
