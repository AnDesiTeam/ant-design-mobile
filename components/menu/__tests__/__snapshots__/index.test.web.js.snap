// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu level 1 renders correctly 1`] = `
<div
  class="am-flexbox am-flexbox-dir-column am-flexbox-align-stretch am-menu"
  style="height:0px;"
>
  <div
    class="am-flexbox am-flexbox-align-top am-menu-select-container"
  >
    <div
      aria-hidden="false"
      class="am-flexbox-item am-menu-select-container-submenu"
      role="tabpanel"
    >
      <div
        class="am-list am-sub-menu"
        style="padding-top:0;"
      >
        <div
          class="am-list-body"
        >
          <div
            class="am-radio-item am-list-item am-list-item-middle"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item1
              </div>
              <div
                class="am-list-extra"
              >
                <label
                  class="am-radio-wrapper"
                >
                  <span
                    class="am-radio"
                  >
                    <input
                      class="am-radio-input"
                      type="radio"
                    />
                    <span
                      class="am-radio-inner"
                    />
                  </span>
                </label>
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
          <div
            class="am-radio-item am-list-item am-list-item-middle"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item2
              </div>
              <div
                class="am-list-extra"
              >
                <label
                  class="am-radio-wrapper"
                >
                  <span
                    class="am-radio"
                  >
                    <input
                      class="am-radio-input"
                      type="radio"
                    />
                    <span
                      class="am-radio-inner"
                    />
                  </span>
                </label>
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Menu multiple select menu level 1 renders correctly 1`] = `
<div
  class="am-flexbox am-flexbox-dir-column am-flexbox-align-stretch am-menu"
  style="height:0px;"
>
  <div
    class="am-flexbox am-flexbox-align-top am-menu-select-container"
  >
    <div
      aria-hidden="false"
      class="am-flexbox-item am-menu-select-container-submenu"
      role="tabpanel"
    >
      <div
        class="am-list am-sub-menu"
        style="padding-top:0;"
      >
        <div
          class="am-list-body"
        >
          <div
            class="am-radio-item am-list-item am-list-item-middle"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item1
              </div>
              <div
                class="am-list-extra"
              >
                <label
                  class="am-checkbox-wrapper"
                >
                  <span
                    class="am-checkbox"
                  >
                    <input
                      class="am-checkbox-input"
                      type="checkbox"
                    />
                    <span
                      class="am-checkbox-inner"
                    />
                  </span>
                </label>
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
          <div
            class="am-radio-item am-list-item am-list-item-middle"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item2
              </div>
              <div
                class="am-list-extra"
              >
                <label
                  class="am-checkbox-wrapper"
                >
                  <span
                    class="am-checkbox"
                  >
                    <input
                      class="am-checkbox-input"
                      type="checkbox"
                    />
                    <span
                      class="am-checkbox-inner"
                    />
                  </span>
                </label>
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="am-multi-select-btns"
  >
    <a
      aria-disabled="false"
      class="am-multi-select-btns-btn am-button am-button-inline"
      role="button"
    >
      <span>
        取 消
      </span>
    </a>
    <a
      aria-disabled="false"
      class="am-multi-select-btns-btn am-button am-button-primary am-button-inline"
      role="button"
    >
      <span>
        确 定
      </span>
    </a>
  </div>
</div>
`;

exports[`Menu multiple select menu renders correctly 1`] = `
<div
  class="am-flexbox am-flexbox-dir-column am-flexbox-align-stretch am-menu"
  style="height:0px;"
>
  <div
    class="am-flexbox am-flexbox-align-top am-menu-select-container"
  >
    <div
      class="am-flexbox-item"
    >
      <div
        class="am-list"
        role="tablist"
      >
        <div
          class="am-list-body"
        >
          <div
            aria-selected="true"
            class="am-menu-selected am-list-item am-list-item-middle"
            role="tab"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item1
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
          <div
            aria-selected="false"
            class="am-list-item am-list-item-middle"
            role="tab"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item2
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="false"
      class="am-flexbox-item am-menu-select-container-submenu"
      role="tabpanel"
    >
      <div
        class="am-list am-sub-menu"
        style="padding-top:0;"
      >
        <div
          class="am-list-body"
        />
      </div>
    </div>
  </div>
  <div
    class="am-multi-select-btns"
  >
    <a
      aria-disabled="false"
      class="am-multi-select-btns-btn am-button am-button-inline"
      role="button"
    >
      <span>
        取 消
      </span>
    </a>
    <a
      aria-disabled="false"
      class="am-multi-select-btns-btn am-button am-button-primary am-button-inline"
      role="button"
    >
      <span>
        确 定
      </span>
    </a>
  </div>
</div>
`;

exports[`Menu renders correctly 1`] = `
<div
  class="am-flexbox am-flexbox-dir-column am-flexbox-align-stretch am-menu"
  style="height:0px;"
>
  <div
    class="am-flexbox am-flexbox-align-top am-menu-select-container"
  >
    <div
      class="am-flexbox-item"
    >
      <div
        class="am-list"
        role="tablist"
      >
        <div
          class="am-list-body"
        >
          <div
            aria-selected="true"
            class="am-menu-selected am-list-item am-list-item-middle"
            role="tab"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item1
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
          <div
            aria-selected="false"
            class="am-list-item am-list-item-middle"
            role="tab"
          >
            <div
              class="am-list-line"
            >
              <div
                class="am-list-content"
              >
                Item2
              </div>
            </div>
            <div
              class="am-list-ripple"
              style="display:none;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="false"
      class="am-flexbox-item am-menu-select-container-submenu"
      role="tabpanel"
    >
      <div
        class="am-list am-sub-menu"
        style="padding-top:0;"
      >
        <div
          class="am-list-body"
        />
      </div>
    </div>
  </div>
</div>
`;
